src/Microsoft.AspNet.Server.Kestrel/Infrastructure/KestrelThread.cs

```diff
@@ -67,10 +67,24 @@ public void Stop(TimeSpan timeout)
             Post(OnStop, null);
             if (!_thread.Join((int)timeout.TotalMilliseconds))
             {
-                Post(OnStopRude, null);
-                if (!_thread.Join((int)timeout.TotalMilliseconds))
+                try
+                {
+                    Post(OnStopRude, null);
+                    if (!_thread.Join((int)timeout.TotalMilliseconds))
+                    {
+                        Post(OnStopImmediate, null);
+                        if (!_thread.Join((int)timeout.TotalMilliseconds))
+                        {
+#if NET451
+                        _thread.Abort();
+#endif
+                        }
+                    }
+                }
+                catch (ObjectDisposedException)
```

Comment:
I'm definitely not a fan of catching an ODE as part of a "normal" shutdown, but I think it is unavoidable since Post relies on the thing we are trying to tear down to still be running.

Since `Thread.Join` failed there's a pretty good chance the loop is still running, but there is no easy way I that I've thought of yet to guarantee that.

Maybe we could add some sort of dispose lock?